node/network/src/proxy/mod.rs

```diff
@@ -0,0 +1,71 @@
+use std::sync::Arc;
+
+use futures::stream::Stream;
+use futures::sync::mpsc::{channel, Receiver, Sender};
+
+use primitives::serialize::Encode;
+
+use crate::peer::PeerMessage;
+use crate::protocol::{forward_msg, PackedMessage};
+
+pub mod dropout;
+
+pub struct Proxy {
+    handlers: Vec<Arc<ProxyHandler>>,
+}
+
+/// All messages before being sent to other participants will be processed by several handlers
+/// chained sequentially in a pipeline.
+///
+/// Each handler will receive an iterator of messages and will return (not necessarily the same)
+/// iterator of messages. This method allows easy manipulation of messages before being emitted
+/// to other authorities through the network.
+///
+/// The order in which handlers are grouped can affect the behavior of the proxy.
+impl Proxy {
+    pub fn new(handlers: Vec<Arc<ProxyHandler>>) -> Self {
+        Self {
+            handlers,
+        }
+    }
+
+    pub fn spawn(&mut self, inc_messages: Receiver<PackedMessage>) {
+        let mut final_messages = Some(inc_messages);
+
+        for mut handler in self.handlers.iter() {
+            let (inc_proxy_message, out_proxy_message) = channel(1024);
```

Comment:
Creating a channel each time we receive a message might be expensive. Besides, if we are doing a broadcasting we would want to dropout the broadcasted message partially and not just to all peers. Also, this code does not mock `forward_msgs`. I suggest `PackedMessage` to be an enum:
```
enum PackedMessage {
  SingleMessage(message, channel),
  BroadcastedMessage(message, channels),
  MultipleMessages(messages, channel)
}

impl PackedMessage {
   fn to_stream(self) -> Box<Stream<Item=(Message, Sender), Error=()>> {
     match self {
       SingleMessage(message, channel) => Box::new(stream::once(Ok((message, channel)))),
       BroadcastedMessage(message, channels) => Box::new(stream::iter_ok(channels.map(|c| (message.clone(), c))),
       MultipleMessages(messages, channel) => Box::new(stream::iter_ok(messages.map(|m| (m, channel.clone()))),
     }
   }
}
```
To avoid copying the message we can use `Arc<Message>`, and create `PackedMessage` like this: `PackedMessage::BroadcastedMessage(Arc::new(m), c)`.

Then each `ProxyHandler` should implement `pipe_stream(&self, s: impl Stream) -> impl Stream`, which would however require from each `ProxyHandler` to store their internal state in `Arc<RwLock<>>`. For example `Dropout` would implement:

```
fn pipe_stream(&self, s: impl Stream) -> impl Stream {
  s.filter(|_|   rng.gen::<f64>() >= self.dropout_rate)
}
```

Then inside `Proxy::spawn` we would do:
```
pub fn spawn(&mut self, inc_messages: Receiver<PackedMessage>) {
let mut prev_stream = Box::new(inc_messages.map(|pm| pm.to_stream()).flatten());
for h in self.handlers {
 prev_stream = Box::new(h.pipe_stream(prev_stream));
}
let task = prev_stream.for_each(|(message, channel)| 
channel.send(message).map(|_|()).map_err(|_| ())
).map(|_| ()).map_err(|_| ());
tokio::spawn(task);
}
```

Reply:
done!